### PR TITLE
Fm facing interface [feature flagged]

### PR DESCRIPTION
**Pull request by nickpell**
Just behind a feature flag, waiting on UAT/QA before feature flag can be removed, but I'd like to get this on master.

Spec: https://docs.google.com/a/dimagi.com/document/d/1HyEB_hRaekL80wKKNT15LFJV3KTnPwkkNpfVv5aAP5Y/edit#

Mockups:
https://confluence.dimagi.com/display/internal/Mockup

@benrudolph @orangejenny 
screen shot 2015-05-03 at 8 16 08 pm

screen shot 2015-05-03 at 8 16 21 pm

### CODE REVIEW

File: corehq/apps/domain/forms.py [hunk 1]
```diff
@@ -1157,3 +1167,430 @@ def process_submission(self, domain=None):
...
+                    date_end=self.cleaned_data['end_date'],
+                    web_user=self.web_user,
+                )
+            else:
+                new_subscription = self.current_subscription.change_plan(
+                    new_plan_version,
+                    date_end=self.cleaned_data['end_date'],
+                    web_user=self.web_user,
+                )
+            if new_subscription.date_start <= datetime.date.today() and datetime.date.today() < new_subscription.date_end:
```
Comment (dimagimon) [inline corehq/apps/domain/forms.py]: E501 line too long (122 > 115 characters)

File: corehq/apps/domain/forms.py [hunk 2]
```diff
@@ -1157,3 +1167,430 @@ def process_submission(self, domain=None):
...
+                )),
+                self.form_actions
+            )
+        else:
+            self.fields['fogbugz_client_name'].initial = self.current_subscription.account.name
+            self.fields['emails'].initial = self.current_subscription.account.billingcontactinfo.emails
+            self.fields['end_date'].initial = self.current_subscription.date_end
+            self.helper.layout = crispy.Layout(
+                TextField('software_plan_edition', plan_edition),
+                crispy.Hidden('software_plan_edition', plan_edition),
```
Comment (orangejenny) [inline corehq/apps/domain/forms.py]: Why is the plan both a text field and a hidden field?

File: corehq/apps/domain/forms.py [hunk 3]
```diff
@@ -1157,3 +1167,430 @@ def process_submission(self, domain=None):
...
+                    '<p><strong>Are you sure you want to extend the subscription?</strong></p>'
+                    '<p>If this project is becoming a self-service project and only paying for '
+                    'hosting fees, please have them self-subscribe through the subscription page.  '
+                    'Please use this page only to extend the existing services contract.</p>'
+                    '</div>'
+                )),
+                self.form_actions
+            )
+
+    def process_subscription_management(self):
```
Comment (orangejenny) [inline corehq/apps/domain/forms.py]: a comment summarizing what this method is for would be helpful

File: corehq/apps/domain/forms.py [hunk 4]
```diff
@@ -1157,3 +1167,430 @@ def process_submission(self, domain=None):
...
+                crispy.Field('emails', css_class='input-xxlarge'),
+                crispy.Field('start_date', css_class='date-picker'),
+                crispy.Field('end_date', css_class='date-picker'),
+                crispy.Field('sms_credits'),
+                crispy.Field('user_credits'),
+                crispy.HTML(_(
+                    '<p><i class="icon-info-sign"></i> Clicking "Update" will set '
+                    'up the subscription in CommCareHQ to one of our standard '
+                    'contracted plans.  If you need to set up a non-standard plan, '
+                    'please email accounts@dimagi.com.</p>'
```
Comment (orangejenny) [inline corehq/apps/domain/forms.py]: should this use `settings.ACCOUNTS_EMAIL`?

File: corehq/apps/domain/templates/domain/internal_subscription_management.html [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+{% block js-inline %}{{ block.super }}
+    <script type="text/javascript">
+        var viewModel = {
+            subscriptionType: ko.observable($('#id_subscription_type').val() || null)
+        };
+        ko.applyBindings(viewModel, document.getElementById('subscription_management'));
+    </script>
+
+    <script>
```
Comment (orangejenny) [inline corehq/apps/domain/templates/domain/internal_subscription_management.html]: No `type="text/javascript"`? same for script block below

File: corehq/apps/domain/forms.py [hunk 6]
```diff
@@ -1157,3 +1167,430 @@ def process_submission(self, domain=None):
...
+                    new_plan_version,
+                    date_end=self.cleaned_data['end_date'],
+                    web_user=self.web_user,
+                )
+            if new_subscription.date_start <= datetime.date.today() and datetime.date.today() < new_subscription.date_end:
+                new_subscription.is_active = True
+            new_subscription.do_not_invoice = False
+            new_subscription.auto_generate_credits = True
+            new_subscription.save()
+        except:
```
Comment (benrudolph) [inline corehq/apps/domain/forms.py]: Not sure if it makes sense here, but would be nice to either log an error so we know what went wrong or display a message to the user

File: corehq/apps/domain/forms.py [hunk 7]
```diff
@@ -1157,3 +1167,430 @@ def process_submission(self, domain=None):
...
+    @property
+    def account_name(self):
+        return self.cleaned_data['fogbugz_client_name']
+
+    @property
+    def account_emails(self):
+        return self.cleaned_data['emails']
+
+    def clean_end_date(self):
+        end_date = self.cleaned_data['end_date']
```
Comment (benrudolph) [inline corehq/apps/domain/forms.py]: I know there was a bug before about `end_date` being None and then trying to compare `None` to a datetime. I saw that it wasn't a required field. let me know if this is something that could potentially happen

File: corehq/apps/domain/templates/domain/internal_subscription_management.html (hunk 5 above)
Comment (benrudolph) [inline corehq/apps/domain/templates/domain/internal_subscription_management.html]: this is actually a noop in html5: http://stackoverflow.com/questions/14323376/do-html5-script-tag-need-type-javascript

Comment (benrudolph) [inline corehq/apps/domain/templates/domain/internal_subscription_management.html]: there are a lot of open script close script tags. maybe we can just make it one?

**Comment by benrudolph** [comment]
none of the comments are blocking i don't think, especially since this is behind a feature flag. would be great to also see a couple of tests if you find any time. merging

File: corehq/apps/domain/forms.py (hunk 7 above)
Comment (nickpell) [inline corehq/apps/domain/forms.py]: good catch, this got fixed in QA

File: corehq/apps/domain/forms.py (hunk 2 above)
Comment (nickpell) [inline corehq/apps/domain/forms.py]: ```TextField``` is just text on the screen, the ```Hidden``` field enables that data to be sent in the form submission

File: corehq/apps/domain/forms.py (hunk 6 above)
Comment (nickpell) [inline corehq/apps/domain/forms.py]: we'll get notified in an email. It'll require dev/ops work to fix the issue (I have in mind conflicting subscriptions).  I agree an error message would be preferable to a 500 page ... but at the moment I'm willing to punt until this is actually observed in the wild.

Comment (benrudolph) [inline corehq/apps/domain/forms.py]: sounds good

**Pull request merged**